packages/plugins/apollo-usage-report/src/index.ts

```diff
@@ -0,0 +1,172 @@
+   */
+  apiKey?: string;
+  /**
+   * Usage report endpoint
+   *
+   * Defaults to GraphOS endpoint (https://usage-reporting.api.apollographql.com/api/ingress/traces)
+   */
+  endpoint?: string;
+};
+
+const DEFAULT_REPORTING_ENDPOINT =
+  'https://usage-reporting.api.apollographql.com/api/ingress/traces';
+
+export function useApolloUsageReport(options: ApolloUsageReportOptions = {}): Plugin {
+  const [instrumentation, ctxForReq] = useApolloInstrumentation(options) as [
+    Plugin,
+    WeakMap<
+      Request,
+      ApolloInlineTraceContext & {
+        schemaId?: string;
+        operationKey?: string;
+      }
+    >,
+  ];
+
+  const {
+    graphRef = process.env['APOLLO_GRAPH_REF'],
+    apiKey = process.env['APOLLO_KEY'],
+    endpoint = DEFAULT_REPORTING_ENDPOINT,
+  } = options;
+
+  let logger: YogaLogger;
+  let schemaId: string;
+
+  return {
+    onPluginInit({ addPlugin }) {
+      addPlugin(instrumentation);
+      addPlugin({
+        async onYogaInit(args) {
+          logger = Object.fromEntries(
+            Object.entries(args.yoga.logger).map(([level, log]) => [
+              level,
+              (...args: unknown[]) => log('[ApolloUsageReport]', ...args),
+            ]),
+          ) as YogaLogger;
+
+          if (!apiKey) {
+            throw new Error(
+              `[ApolloUsageReport] Missing API key. Please provide one in plugin options or with 'APOLLO_KEY' environment variable.`,
+            );
+          }
+
+          if (!graphRef) {
+            throw new Error(
+              `[ApolloUsageReport] Missing Graph Ref. Please provide one in plugin options or with 'APOLLO_GRAPH_REF' environment variable.`,
+            );
+          }
+
+          logger.debug('using', { apiKey, graphRef });
+        },
+        async onSchemaChange({ schema }) {
+          if (schema) {
+            schemaId = await hashSHA256(printSchema(schema));
+          }
+        },
+        onParse() {
+          return ({ result, context: { request, params } }) => {
+            const ctx = ctxForReq.get(request);
+            if (!ctx) {
+              logger.debug('ctx not found during parsing');
+              return;
+            }
+
+            const { operationName } = params;
+            const signature = defaultUsageReportingSignature(result, operationName || '');
+            ctx.operationKey = `# ${operationName || '-'}\n${signature}`;
+            ctx.schemaId = schemaId;
+          };
+        },
+
+        async onResultProcess(args) {
+          const ctx = ctxForReq.get(args.request);
+          if (!ctx?.operationKey || !ctx?.schemaId) {
+            logger.debug('ctx not found during result processing:', ctx);
+            return;
+          }
+
+          // TODO: Handle async iterables ?
+          if (isAsyncIterable(args.result)) {
+            logger.debug('async iterable results not implemented for now');
+            return;
+          }
+
+          fetch(endpoint, {
```

Comment:
```suggestion
          args.ctx.waitUntil(fetch(endpoint, {
```
Let's use `waitUntil` to put this in the queue